src/components/FlatTreeRenderer.js

```diff
@@ -0,0 +1,436 @@
+        yOffsetObject: {
+          ...acc.yOffsetObject,
+          [item.key]: offset,
+        },
+      }
+    },
+    { depthTableArray: [], offsetAbove: 0, yOffsetObject: {} }
+  )
+
+  return yOffsetObject
+}
+
+/**
+ *
+ */
+const calculateXOffset = (item, visibleStartDepth) => {
+  const isFirstColumn = item.viewInfo.table.column === 1
+  const isSecondColumn = item.viewInfo.table.column === 2
+
+  const depth = item.path.length - visibleStartDepth
+
+  const xOffsetCount =
+          (depth * DEPTH_OFFSET +
+            (isFirstColumn ? TABLE_FIRST_COLUMN_OFFSET : 0) +
+            // for table view second column
+            (isSecondColumn
+              ? TABLE_SECOND_COLUMN_OFFSET + TABLE_FIRST_COLUMN_WIDTH
+              : 0) +
+            // deeper nodes adjust offset for the number of active first and second table columns above them
+            item.viewInfo.table.tableSecondColumnsAbove *
+              TABLE_FIRST_COLUMN_WIDTH) *
+          1.2
+
+  return xOffsetCount
+}
+
+/**
+ * Takes flat array and returns y offset and a function that updates height.
+ * On setting updated height object using the `setHeightObject` calculates new yoffset and returns updated yOffset.
+ */
+const useYoffsetHandler = flatArray => {
+
+  // only re-render for change in yOffsetObject
+  const heightObjectRef = useRef({})
+  const flatArrayRef = useRef(flatArray)
+
+  // we don't want setHeightObject to rebuild and cause re-render. so using ref to pass updated dependency
+  flatArrayRef.current = flatArray
+  const [yOffsetObject, setYO] = useState({})
+
+  // side-effect to update y offset when flatArray changes
+  useEffect(() => {
+    setHeightObject({})
+  }, [flatArray])
+
+  /** This height takes updated height obj and calculates y offset obj. */
+  const setHeightObject = useCallback(updatedHeightObject => {
+    heightObjectRef.current = { ...heightObjectRef.current, ...updatedHeightObject }
+    const updatedOffsetObj = calculateYoffset(heightObjectRef.current, flatArrayRef.current)
+    setYO(updatedOffsetObj)
+  }, [])
+
+  return { yOffsetObject, heightObj: heightObjectRef, setHeightObject }
+}
+
+/**
+ * Component that handles flat tree group animations.
+ */
+const TreeAnimation = ({
+  flatArray,
+  flatArrayKey,
+  visibleStartDepth,
+  oldFlatArray,
+  oldFlatArrayKey
+}) => {
+  const { yOffsetObject, heightObj, setHeightObject } = useYoffsetHandler(flatArray)
+  const yOffsetObjectOldRef = useRef(yOffsetObject)
+  const heightObjectPendingRef = useRef([])
+
+  useEffect(() => {
+    yOffsetObjectOldRef.current = yOffsetObject
+  }, [yOffsetObject])
+
+  const yOffsetObjectOld = yOffsetObjectOldRef.current
+
+  const transitions = useTransition(
+    flatArray,
+    node => node.key,
+    {
+      config: SPRING_CONFIG_GROUP,
+      unique: true,
+      from: {
+        opacity: 0
+      },
+      enter: i => async next => {
```

Comment:
By convention `i` always refers to a `number` within a loop. I suggest calling this `transitionItem` since `item` is used.